### ISSUE TITLE
x-kubernetes-map-type=atomic is required if a containing array has x-kubernetes-list-type=set

### ISSUE THREAD

**Issue by tatsuhiro-t** [kubernetes-sigs/controller-tools#752]
I encountered validation error when applying the generated CRD containing PodSpec which has new [ResourceClaim](https://pkg.go.dev/k8s.io/api@v0.26.0/core/v1#ResourceClaim) in https://pkg.go.dev/k8s.io/api@v0.26.0/core/v1#ResourceRequirements added in k8s v1.26.
It sets `+listType=set`.

When applying generated CRD, I got the following error:
> ... .properties[spec].properties[initContainers].items.properties[resources].properties[claims].items.x-kubernetes-map-type: Invalid value: "null": must be atomic as item of a list with x-kubernetes-list-type=set

This validation seems to be happened at https://github.com/kubernetes/apiextensions-apiserver/blob/9746985976f9fee21736250e3afb0eae27026633/pkg/apis/apiextensions/validation/validation.go#L945

Adding `x-kubernetes-map-type=atomic` under items fixes this issue.

I'm wondering whether this is a lack of `x-kubernetes-map-type=atomic` in ResourceClaim, or something that CRD generator must deal with.

The following patch automatically adds `x-kubernetes-map-type=atomic` to items when `x-kubernetes-list-type=set` is seen, but I feel this is hacky and probably not a correct way:

```go
--- a/pkg/crd/markers/topology.go
+++ b/pkg/crd/markers/topology.go
@@ -116,6 +116,14 @@ func (l ListType) ApplyToSchema(schema *apiext.JSONSchemaProps) error {
        }
        p := string(l)
        schema.XListType = &p
+
+       if l == "set" {
+               if itemSchema := schema.Items.Schema; itemSchema != nil {
+                       v := "atomic"
+                       itemSchema.XMapType = &v
+               }
+       }
+
        return nil
 }

```

**Comment by kluevandrew** [comment]
Same issue for me, any workaround?

**Comment by kluevandrew** [comment]
> Same issue for me, any workaround?

I found a workaround by myself, **modify according  to your resources**

```diff
# ./config/crd/kustomization.yaml

  # the following config is for teaching kustomize how to do kustomization for CRDs.
  configurations:
    - kustomizeconfig.yaml
+
+ patchesJson6902:
+  - path: patches/workaround_in_roles.yaml
+   target:
+      group: apiextensions.k8s.io
+      version: v1
+      kind: CustomResourceDefinition
+      name: roles.tarantool.io
```

```diff
# ./config/crd/patches/workaround_in_roles.yaml

+ - op: add
+   path: /spec/versions/0/schema/openAPIV3Schema/properties/spec/properties/storageTemplate/properties/spec/properties/template/properties/spec/properties/ephemeralContainers/items/properties/resources/properties/claims/items/x-kubernetes-map-type
+   value: atomic
+ - op: add
+   path: /spec/versions/0/schema/openAPIV3Schema/properties/spec/properties/storageTemplate/properties/spec/properties/template/properties/spec/properties/containers/items/properties/resources/properties/claims/items/x-kubernetes-map-type
+   value: atomic
+ - op: add
+   path: /spec/versions/0/schema/openAPIV3Schema/properties/spec/properties/storageTemplate/properties/spec/properties/template/properties/spec/properties/initContainers/items/properties/resources/properties/claims/items/x-kubernetes-map-type
+   value: atomic
+ - op: add
+   path: /spec/versions/0/schema/openAPIV3Schema/properties/spec/properties/storageTemplate/properties/spec/properties/template/properties/spec/properties/volumes/items/properties/ephemeral/properties/volumeClaimTemplate/properties/spec/properties/resources/properties/claims/items/x-kubernetes-map-type
+   value: atomic
+ - op: add
+   path: /spec/versions/0/schema/openAPIV3Schema/properties/spec/properties/storageTemplate/properties/spec/properties/volumeClaimTemplates/items/properties/spec/properties/resources/properties/claims/items/x-kubernetes-map-type
+   value: atomic
+ - op: add
+   path: /spec/versions/1/schema/openAPIV3Schema/properties/spec/properties/replicasetTemplate/properties/volumeClaimTemplates/items/properties/spec/properties/resources/properties/claims/items/x-kubernetes-map-type
+   value: atomic
+ - op: add
+   path: /spec/versions/1/schema/openAPIV3Schema/properties/spec/properties/replicasetTemplate/properties/podTemplate/properties/spec/properties/containers/items/properties/resources/properties/claims/items/x-kubernetes-map-type
+   value: atomic
+ - op: add
+   path: /spec/versions/1/schema/openAPIV3Schema/properties/spec/properties/replicasetTemplate/properties/podTemplate/properties/spec/properties/ephemeralContainers/items/properties/resources/properties/claims/items/x-kubernetes-map-type
+   value: atomic
+ - op: add
+   path: /spec/versions/1/schema/openAPIV3Schema/properties/spec/properties/replicasetTemplate/properties/podTemplate/properties/spec/properties/volumes/items/properties/ephemeral/properties/volumeClaimTemplate/properties/spec/properties/resources/properties/claims/items/x-kubernetes-map-type
+   value: atomic
+ - op: add
+   path: /spec/versions/1/schema/openAPIV3Schema/properties/spec/properties/replicasetTemplate/properties/podTemplate/properties/spec/properties/initContainers/items/properties/resources/properties/claims/items/x-kubernetes-map-type
+   value: atomic
+ 
```

**Comment by kluevandrew** [comment]
Is it still not fixed as i can see, 0.11.1 has a revert commit of fix. I think you need to re-open that issue.

**Comment by kluevandrew** [comment]
Found discussion here https://github.com/kubernetes-sigs/controller-tools/pull/753
